### ISSUE TITLE
Support specifying type of "this" in function implementations

### ISSUE THREAD

**Issue by abergs** [Microsoft/TypeScript#229]
Migrated issue from codeplex: https://typescript.codeplex.com/workitem/507

Currently typescript will type the "this" pointer in function callbacks as "any." Arrow syntax lets us capture "this" from the outer scope, and this can enable proper typing. It would be nice to be able to provide
an optional "ambient this" declaration in function signatures:

    (declare this : MyType, first: number, second: string) : any;

The rules would be:

1) the declaration is obviously optional
2) if specified, should be first parameter
3) cannot be used with arrow syntax lambdas (should be compiler error)

I know you can cast as needed to get the intellisense but I'd rather not put this responsibility on the implementer of the function. Ideally it should be part of the definition.

Example with Sammy:

    var app = $.sammy("#view", function() {
         // define default route
         this.get("#/", function() { ... } 
    });
    

In the above case, typescript can only type this as any.

Thoughts?

**Comment by hraban** [comment]
Second @pspeter3's comment: this would be great for the React definitions.

@cspotcode: the proposal looks intuitive. wrt implicit this: it should definitely be default on methods, eventually. Since it's backwards incompatible, perhaps default --noImplicitThis flag on (so feature off) for now, and default it to off in the next (major) version?

Implicit this on methods is a strong feature that deserves to be enabled by default soon, so if the next major rev is too far off, perhaps use the next minor (1.5?) as a beta period (default off) and default on at 1.6?

Lack of implicit method binding in JS is a common source of confusion for newcomers, and TS makes this even less obvious. Foo.prototype.meth = func() gives you a hint that it's not common, but class Foo { meth() { ...} }, well, you have to really understand JS and how TS compiles to JS to see why ~~`((new Foo()).meth)()`~~  `var f = (new Foo()).meth; f();` won't work. There, I just got confused myself... I would've expected `((new Foo()).meth)()` to use caller's `this`, but it's bound to that fresh Foo instance.

All in all, definitely plus one from me. I'd enable it asap.

**Comment by aholmes** [comment]
Regarding a "noImplicitThis" flag, I propose not adding this and learning from PHP's mistakes. These kinds of flags caused all sorts of problems between different versions of PHP and different libraries using code that "worked" with different flags enabled than what the current system had enabled.

If implicit this is added and is a BC break I propose that this change is made prominent so any developers updating TS are made aware of this.

**Comment by nathggns** [comment]
Any update on what's happening with this?

**Comment by cspotcode** [comment]
As far as I know, the Typescript team is still waiting for someone to write a design proposal, which @RyanCavanaugh says in this comment: https://github.com/Microsoft/TypeScript/issues/229#issuecomment-50406084

See also: the guidelines for writing a proposal: https://github.com/Microsoft/TypeScript/wiki/Writing-Good-Design-Proposals

**Comment by nathggns** [comment]
@cspotcode Gotcha. Submitted a proposal over at #1985. It's mostly based on your work so would appreciate the feedback.

**Comment by cspotcode** [comment]
Very nice, I will take a look soon.  Thanks!

**Comment by aclave1** [comment]
For the time being, can we just use?

```
var self:MyClassType = this;
```

**Comment by basarat** [comment]
@aclave1 yup. But then its a responsibility of the API user to remember to do this *every single time*. Becomes annoying fairly fast.

**Comment by nycdotnet** [comment]
I will second that this is the workaround I use *and* that it gets annoying really fast.  My suggestion to formalize the ability to use a capture variable (such as if you've converted a JS function to a TS class and don't want to avoid the risky busy work the compiler currently mandates):  https://github.com/Microsoft/TypeScript/issues/1859

**Comment by aclave1** [comment]
@basarat @nycdotnet I agree with you guys.

**Comment by englercj** [comment]
Consider also the case where you can specify a callback and the context it will be called with. For example, an EventEmitter3 event handler:

```js
emitter.on('somevent', function () {
    // this here is basically 'any', but I want it to be a specific type that I can specify with an annotation
    // without having to do a function bind or a self scope var.
}, this);
```

**Comment by redexp** [comment]
Hey guys. Looks like we will never figure out how statically (I mean .d.ts files) type `this` scope. What about to switch from config to code? Just like grunt (with tons of static configs) to gulp (with nice modular "config" in code). For example we will have *function this type callback* witch should return type of *this*.
```javascript
function getOnMethodThisType(on) {
  if (on.arguments.length === 3) {
    return on.arguments[2].type;
  }

  return TS.globalScope.type;
}
```
Thinking about "code instead of config", maybe we need new definition file type where instead of pseudo code config will be just js/ts code? For example
```typescript
TS.class('Emitter')
  .method('on', function (on: TSMethod) {
    if (!on.callee) {
      throw new TS.Error.CalleeRequired();
    }

    switch (on.arguments.length) {
      case 2:
        return {
          "arguments": [String, Function],
          "this": TS.global.type,
          "return": on.callee.type
        };
      case 3:
        return {
          "arguments": [String, Function, Any],
          "this": on.arguments.type,
          "return": on.callee.type
        };
    }

    throw new TS.Error.WrongArguments();
  });
```
Pros: modular code, share code through npm, custom errors like "One argument is deprecated since v1.4", code is more readable than `.d.ts`, will solve any `this` problems.

**Comment by lazutkin** [comment]
:+1:

**Comment by lucasmciruzzi** [comment]
:+1: ... right now I'm using the "self workaround", and I'm not enjoying it Dx

**Comment by JsonFreeman** [comment]
I agree with @ivogabe at https://github.com/Microsoft/TypeScript/issues/229#issuecomment-53194534. If it is a type parameter, that means the type of 'this' can vary on a per-call basis. And if it varies, then a 'this' expression will not have a known type in the body. Better to have it be a parameter with a type annotation.

**Comment by 304NotModified** [comment]
:+1:

**Comment by impinball** [comment]
FWIW, this is required for a lot of .d.ts files on DefinitelyTyped. Here's an incomplete list of modules that require this:

- `jquery`
- `underscore`
- `lodash`
- `lazy.js`

Also, many of the native ECMAScript APIs use bound `this` arguments, such as many of the `Array.prototype` methods and some of the String methods. `JSON.{stringify,parse}` call their callback with either an object or array.

```ts
// Array generics
interface Callback<T, U> {
  (this: T, value: U, index: number, array: ArrayLike<U>): any;
}

interface ArrayLike<T> {
  [key: number]: T;
  length: number;
}

interface Array<T> {
  // things...
  forEach(this: ArrayLike<T>, callback: Callback<ArrayLike<T>, T>): void;
  forEach<U>(this: ArrayLike<T>, callback: Callback<U, T>, thisArg: U): void;
  // more things...
}

// JSON
interface JSONCallback {
  (this: any[], key: number, value: any) => any;
  (this: {[key: string]: any}, key: string, value: any) => any;
}

interface JSON {
  parse(text: string, reviver?: JSONCallback): any;
  stringify(value: any): string;
  stringify(value: any, reviver?: JSONCallback | string[], space?: string | number): string;
}

// Function methods
interface Function {
  bind<T, U>(thisArg: T, ...args: any[]): (this: T, ...args: any[]) => U;
}
```

**Comment by kobezzza** [comment]
+1 Very need this feature.

**Comment by eggers** [comment]
+1

Also voting for the syntax suggested by @Bartvds

```typescript
interface Widget {
    sprock<this:SomeClass>(): void;
}
```

over

```typescript
interface Widget {
    sprock(this:SomeClass): void;
}
```

**Comment by mhegazy** [comment]
This is now tracked by #3694.

**Comment by Zorgatone** [comment]
+1

**Comment by distinctdan** [comment]
As an easy workaround, you can declare a variable at the top of the function, give it a type, and set it equal to `this`. This allows you to do type checking on it, as long as you don't mind renaming "this":

```myMethod(a: number) {
    var self: MyClass = this;
    self.doStuff();
}````

**Comment by QueueHammer** [comment]
My work around for the last 2 years has been to just not use TypeScript. Though it's was for other issues at the time, which some of them have been fixed since v0.8.

**Comment by hraban** [comment]
@QueueHammer thanks man, but not terribly constructive. Did you find an alternative that does fix this problem, and how?

**Comment by SaschaNaz** [comment]
This is fixed by #6739 on April 8.

**Comment by kitsonk** [comment]
But isn't released yet (it is in `typescript@next` and will be released in 2.0).

**Comment by ksnyde** [comment]
Has this now been released as part of 2.0? Is there a set of reference examples of using this?

**Comment by kitsonk** [comment]
Yes.  It is sort of covered here: https://www.typescriptlang.org/docs/handbook/functions.html

But taking an example from above:

```typescript
myMethod(this: MyClass) {
    this.doStuff();
}
```
